shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/Serialization.hs

```diff
@@ -137,3 +156,85 @@ rationalFromCBOR = do
     case ints of
       n:d:[] -> pure $ n % d
       _ -> cborError $ DecoderErrorSizeMismatch "rational" 2 numInts
+
+-------------------------------------------------------------------------
+-- Annotated Decoder
+-------------------------------------------------------------------------
+
+-- | An AnnotatedDecoder produces a value which needs a reference to the
+-- original ByteString to be constructed. For example, consider
+--
+-- `data Foo = Foo Int ByteString`
+--
+-- where the ByteString is expected to be the serialized form of Foo.
+--
+-- The pattern `AnnotatedDecoder` takes care that the bytes provided are the
+-- correct ones.
+newtype AnnotatedDecoder s a
+  = AnnotatedDecoder { unwrapAnn :: (Decoder s (LByteString -> a)) }
+  deriving (Functor)
+
+instance Applicative (AnnotatedDecoder s) where
+  pure x = AnnotatedDecoder $ const <$> pure x
+  (AnnotatedDecoder a) <*> (AnnotatedDecoder b) =
+    AnnotatedDecoder $ (<*>) <$> a <*> b
+
+decodeAnnotated :: forall a. (Typeable a , FromCBORAnnotated a)
+  => LByteString
+  -> Either DecoderError a
+decodeAnnotated = decodeAnnotatedDecoder (show . typeRep $ Proxy @a) fromCBORAnnotated
+
+decodeAnnotatedDecoder :: Text -> (forall s. AnnotatedDecoder s a) -> LByteString -> Either DecoderError a
+decodeAnnotatedDecoder label' decoder bytes =
+  (\x -> x bytes) <$> decodeFullDecoder label' (unwrapAnn decoder) bytes
+
+withSlice :: AnnotatedDecoder s (LByteString -> a) -> AnnotatedDecoder s a
+withSlice (AnnotatedDecoder dec) = AnnotatedDecoder $ do
+  (k, start, end) <- decodeWithByteSpan dec
+  return $ \bytes -> k bytes (sliceOffsets start end bytes)
+  where
+  sliceOffsets :: ByteOffset -> ByteOffset -> LByteString -> LByteString
+  sliceOffsets start end = (BSL.take (end - start) . BSL.drop start)
+
+withSlice' :: forall s a. AnnotatedDecoder s (ByteString -> a) -> AnnotatedDecoder s a
+withSlice' = withSlice . fmap (. BSL.toStrict)
+
+-- | Wrap a plain decoder into an annotated one.
+liftAnn :: Decoder s a -> AnnotatedDecoder s a
+liftAnn dec = AnnotatedDecoder $ const <$> dec
+
+-- | Wrap a plain decoder into an annotated one that populates the ByteString with a slice.
+withAnnotationSlice :: forall s a. Decoder s (LByteString -> a) -> AnnotatedDecoder s a
+withAnnotationSlice = withSlice . liftAnn
+
+-- | Strict version of withAnnotationSlice
+withAnnotationSlice' :: forall s a. Decoder s (ByteString -> a) -> AnnotatedDecoder s a
+withAnnotationSlice' = withSlice' . liftAnn
+
+class FromCBORAnnotated a where
+  fromCBORAnnotated :: forall s. AnnotatedDecoder s a
+
+fromCBOREmptyAnnotation :: FromCBORAnnotated a => Decoder s a
+fromCBOREmptyAnnotation = (\x -> x mempty) <$> unwrapAnn fromCBORAnnotated
+
+-------------------------------------------------------------------------
+-- Wrapped Decoder
+-------------------------------------------------------------------------
+
+-- | Wrap both annotated and plain decoders
+data WrappedDecoder a =
```

Comment:
This code is dead, is it not?

Reply:
All the changes in this file are temporary code copied and pasted from the `nc/annotatoin` branch on cardano-base.